internal/resolver/dns/dns_resolver_test.go

```diff
@@ -1215,3 +1216,56 @@ func (s) TestReportError(t *testing.T) {
 		}
 	}
 }
+
+// Override the default dns.ResolvingTimeout with a test duration.
+func overrideResolveTimeoutDuration(t *testing.T, dur time.Duration) {
+	origDur := dns.ResolvingTimeout
+	dnspublic.SetResolvingTimeout(dur)
+
+	t.Cleanup(func() { dnspublic.SetResolvingTimeout(origDur) })
+}
+
+// Test verifies that when the DNS resolver gets timeout error when net.Resolver
+// takes too long to resolve a target.
+func (s) TestResolveTimeout(t *testing.T) {
```

Comment:
I'm not understanding this test now.

I think the test was perfect the way you had it before, but instead of using a `Sleep`, you'd use a blocked channel send instead to cause the error.

I.e. in `func (tr *testNetResolver) LookupHost()`, use `SendContext` instead of `Send` so the function can return an error when the context it receives expires:

```go
if tr.lookupHostCh != nil {
	if err := tr.lookupHostCh.SendContext(ctx, nil); err != nil {
		return err
	}
}
```

Then in this test; mostly like how you had it before:

```go
	// Set DNS resolving timeout to 7ms
	// ...

	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
	defer cancel()

	// ...
	const target = "infinity"
	overrideNetResolver(t, &testNetResolver{
		lookupHostCh:    testutils.NewChannelWithSize(0),
		hostLookupTable: map[string][]string{},
	})

	// Building will produce a timeout error on errCh since hostLookupCh is blocked.
	r, _, errCh := buildResolverWithTestClientConn(t, target)

	select {
	case <-ctx.Done():
		t.Fatal("Timeout when waiting for an error from the resolver")
	case err := <-errCh:
		if err == nil || !strings.Contains(err.Error(), "hostLookup timeout") {
			t.Fatalf(`we expect to see timed out error`)
		}
	}
}
```